### PR TITLE
Fail regress if row estimates changes

### PR DESCRIPTION
**Pull request by dgkimura**
In GPDB, regress has a post-processing step to reformat the result and
expected files. That's used, for example, to allow a test to be agnostic
to row order if ORDER BY is not defined. That's necessary because in an
MPP environment, network ordering is nondeterministic.

One post-processing step strips row estimate information. But that makes it
impossible to test row estimates (e.g. row hints) in regress. Instead, a test
should explicitly omit that information through EXPLAIN parameters (costs=off)
if it is irrelevant.

In GPDB we've been inconsistent with using EXPLAIN (costs=off) because the 
post-processing step made it irrelevant. Ideally, we would update our tests. I
think that can be done incrementally. For now, I added a matchsubs to mask out
the diffs for stale tests.

--

I broke this pull-request into 2 commits:
- b255362e96fe0107de5004d3352e641d4ba4aa12 Perl script post-processing changes
- df7862557a4b196cb3d599155a9affd7a6a741c6 test fix ups

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-fail-regress-row-diffs-rocky8

### CODE REVIEW

File: src/test/isolation2/expected/ao_blkdir.out [hunk 1]
```diff
@@ -1,6 +1,11 @@
 -- White-box tests asserting composition of AO/CO block directory entries.
 -- All tuples are directed to seg0 and each INSERT has an increasing row count
 -- to make their identification easy.
+--
+-- start_matchsubs
+-- m/\(cost=.*\)/
+-- s/\(cost=.*\)//
```
Comment (avamingli) [inline src/test/isolation2/expected/ao_blkdir.out:7]: Is it possible to put them in one file, like init_file ?

**Review by gpopt — APPROVED** [review]
+1

File: src/test/isolation2/expected/ao_blkdir.out (hunk 1 above)
Comment (dgkimura) [reply src/test/isolation2/expected/ao_blkdir.out:7]: Yeah, it's possible.  But ultimately I'd like to add tests that should fail if costs change.  And I'm not sure whether it's possible to turn off the match for a test if the match is put in the init_file.

I think the best final solution is to change our tests to EXPLAIN (costs off).  For now, I'm proposing an interim state with matchsubs until we get there.  Thoughts?

**Review by soumyadeep2007 — APPROVED** [review]
Don't mind the change at all! We should encourage more uses of costs off where we don't care about costs and actually consider costs when we need to (like for hint testing etc).

**Comment by THANATOSLAVA** [comment]
Sounds like going forward unless we have "cost off" the row count diffs would be flagged as errors.

Do we still have the option of not flagging row count diffs but just keep it showing as a reference? Eg. I don't mind small fluctuations in the row count, but do want to see a number to justify a plan shape?

**Comment by dgkimura** [comment]
On Tue, Dec 5, 2023 at 9:04 AM THANATOSLAVA wrote:
> Do we still have the option of not flagging row count diffs but just keep it showing as a reference? Eg. I don't mind small fluctuations in the row count, but do want to see a number to justify a plan shape?

I think there are two options. Either 1) use start/end_ignore or 2) use the matchsubs pattern as proposed as an interim state in this PR.

**Review by chrishajas — APPROVED** [review]
I like this since we can properly test and capture cardinality estimate changes. It does make finding the root cause of "random CI flakes" more difficult when we use `EXPLAIN (COSTS OFF)`, as most of the flakes are due to unanalyzed tables or changes in plan shape with an identical cost.

**Comment by dgkimura** [comment]
Merged. Thanks for the reviews!

**Pull request merged**